### PR TITLE
[MLIR][Vector] Add Lowering for vector.step

### PR DESCRIPTION
**Pull request by manupak**
Currently, the lowering for vector.step lives
under a folder. This is not ideal if we want
to do transformation on it and defer the
 materizaliztion of the constants much later.

This commits adds a rewrite pattern that
could be used by using
`transform.structured.vectorize_children_and_apply_patterns`
transform dialect operation.

Moreover, the rewriter of vector.step is also
now used in -convert-vector-to-llvm pass where
it handles scalable and non-scalable types as
LLVM expects it.

As a consequence of removing the vector.step
lowering as its folder, linalg vectorization
will keep vector.step intact.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp:33]: a lot of headers for such a small rewrite, please could you go through these and remove ones that are unused?

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp:42]: `ArithOps` is a bit ambiguous, perhaps `StepToArithConstantOp`? Also applies to same name used elsewhere

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp:56]: better to use an early-exit here: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code
```suggestion
    if (resultType.isScalable())
      return failure();
      
    SmallVector<APInt> indices;
    for (unsigned i = 0; i < resultType.getNumElements(); i++)
      indices.push_back(APInt(/*width=*/64, i));
    rewriter.replaceOpWithNewOp<arith::ConstantOp>(
        stepOp, DenseElementsAttr::get(resultType, indices));
    return success();
```

**Comment by c-rhodes** [comment]
>This is not ideal if we want
to do transformation on it and defer the
materizaliztion of the constants much later.

makes sense, the fact it's a simple step is lost after the canonicalization and I didn't consider the value in keeping that around when I added the operation

**Comment by manupak** [comment]
cc: @Groverkss

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp:50]: Follow the llvm coding style for `for` loops: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop 
```suggestion
      for (unsigned i = 0, e = resultType.getNumElements(); i != e; ++i)
```

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp:42]:
```suggestion
struct StepToArithOps final : OpRewritePattern<vector::StepOp> {
```

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp:48]: nit: Invert this condition and return early

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (kuhar) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp:50]: Also let's not reallocate every time we reach the capacity; the final vector size is known.
We could do `enumerate` over the vector of the final size or `auto indices = map_to_vector(seq(n), [](int64_t x) { return APInt(64, x); });

**Comment by banach-space** [comment]
Thanks for working on this! I have a few high level comments.

> This is not ideal if we want to do transformation on it and defer the materizaliztion of the constants much later.

Do we? Is there a use case for this? I just want to make that we are not moving code for the sake of ... moving code around 😅 

Re layering, transformations in "mlir/lib/Dialect/Vector/Transforms/LowerVector{.*}.cpp" _lower_ from higher-level Vector Ops to lower-level Vector ops. Rewriting `vector.step` as `arith.constant` feels  more like dialect conversion to me. But then we don't have `VectorToArith`. So, to me, there are actual benefits to keeping this as a folder. 

Re changes in tests, we should avoid adding more patterns and TD Ops to the vectorizer tests that already use `transform.structured.vectorize_children_and_apply_patterns` (this TD Op should encapsulate everything that's needed).

What happens if you don't add `transform.apply_patterns.vector.step_to_arith` to the tests?

**Comment by Groverkss** [comment]
> > Do we? Is there a use case for this? I just want to make that we are not moving code for the sake of ... moving code around 😅

We do actually have a usecase downstream. Sometimes, we see patterns like:

```
%a = vector.step : vector<128xindex>
%b = vector.extract_strided_slice %thread_id, %a : vector<4xindex>
```

If you know it's a vector.step, you can fold the step + slice into:

```
%a = vector.step : vector<4xindex>
%b = arith.constant ...
%c = arith.addi %b, %a
%d = arith.addi %c, %thread_id
```

So you don't have to materialize a big constant like vector<128xindex> into private memory. Having vector.step allows you to know how the arith.constant materializes.

**Comment by manupak** [comment]
Thanks all, I ll be working on addressing the specific comments.

Replying to higher-level comments from @banach-space,

> Do we? Is there a use case for this? I just want to make that we are not moving code for the sake of ... moving code around

Yes, I think @Groverkss is spot on the use case here. 
In a higher-level, `vector.step` carry the semantics of specialized constant which has `f(i) = i` property. This is immediately lost if 'fold' it away.  Therefore I d consider that as a lowering to lower level abstraction over a 'folder'.

> Re layering, transformations in "mlir/lib/Dialect/Vector/Transforms/LowerVector{.*}.cpp" lower from higher-level Vector Ops to lower-level Vector ops

Is this true ?
I can cite few examples that violates the notion that LowerVector*.cpp exclusively contains `vector` -> `vector` rewrites such as e.g. : LowerVectorContract. I suppose arith dialect is a bit special in notion that it inter mixes with most dialects. If we are to push for a seperation of dialects via a dialect conversion, I feel we will have to replicate most operations in `vector` (such as `vector.add`) .. Therefore, Im not fully following the case for a `VectorToArith` because its feels impossible `vector` to be practically used w/o `arith`.

> So, to me, there are actual benefits to keeping this as a folder.

Im not following the logical inference of the absense of  `VectorToArith` to a keep this as a folder. Can you help me understand this a bit more?
At least to me personally, folders should be proven always beneficial to have the transformation where here we are presenting a counter example for the folding of `vector.step`.

> Re changes in tests, we should avoid adding more patterns and TD Ops to the vectorizer tests that already use transform.structured.vectorize_children_and_apply_patterns (this TD Op should encapsulate everything that's needed).

> What happens if you don't add transform.apply_patterns.vector.step_to_arith to the tests?

This part I somewhat agree. 

So there was three options : 
A1) Introduce and use `transform.apply_patterns.vector.step_to_arith` to showcase expectation of tests are still met (more or less) using the lowering. -- I took this one but happy to change
A2) include the rewriter in `transform.structured.vectorize_children_and_apply_patterns`
A3) keep it as `vector.step` 

I can switch to A2 and A3. Having read again what `transform.structured.vectorize_children_and_apply_patterns` entails, I feel A2 is better choice. WDYT ?

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (manupak) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp]: Apologies, too much 'inspiration' from other files :). removed most of it.

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (manupak) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp]: makes sense. changed.

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (manupak) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp]: done

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (manupak) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp]: thanks for the tips. changed now.

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (manupak) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp]: done.

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (manupak) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp]: done as per @c-rhodes comment earlier as well.

File: mlir/include/mlir/Dialect/Vector/Transforms/LoweringPatterns.h
Comment (manupak) [reply mlir/include/mlir/Dialect/Vector/Transforms/LoweringPatterns.h]: done.

File: mlir/include/mlir/Dialect/Vector/Transforms/LoweringPatterns.h
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/Vector/Transforms/LoweringPatterns.h:240]: pattern name needs updating

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp:26]: nit: ->`VectorStepToArithConstantLowering` or similar

**Review by dcaballe — APPROVED** [review]
Thanks! LG!

**Review by c-rhodes — APPROVED** [review]
LGTM cheers

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp
Comment (manupak) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorStep.cpp]: done

**Comment by Groverkss** [comment]
> Cool, why not upstream that? That's a good optimisation and nice justification for this change.

It would eventually be upstreamed. But upstreaming a transformation, which can never actually be used because the op folds itself away, is upstreaming something that cannot get used.

I'm not sure this change by itself needs justification though. In the original review for vector.step, moving this transformation out of canonicalization was already requested but not addressed https://github.com/llvm/llvm-project/pull/96776/files#r1667009554 .

**Comment by manupak** [comment]
Thanks @banach-space for taking another look..

> Can you link a specific example from that file that doesn't lower to other Vector Ops?

> Bare in mind that there are many examples in-tree that e.g. don't follow our coding guidelines. However, that doesn't mean that it's OK to violate them, it merely means that there's code that should be re-formatted ;-)

https://github.com/llvm/llvm-project/blob/340cd4e631d72d02cd79f9aad74d2a354abc977e/mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp#L1374-L1380

I did not mean to cite as "just because there are other code that does the same thing". What I wanted bring up, you would need `arith` dialect more inter-twined with other shaped type dialects such as `vector` and `tensor`. Therefore a dialect conversion that from `VectorToArith` did not make sense to me.

> A2 is better choice.

I can do A2.

> As for testing, I'd add this pattern to TestVectorScanLowering (and rename it as TestVectorToArithLowering). That would be the least intrusive and the most straightforward path for now.

May I ask why `vector.scan` is special to be combined with `vector.step` ? Only reason I see its rewrite pattern is named as "ScanToArithOps" but that seemed like a random choice. For e.g. `vector.mask` do use `arith` ops as well many other ops in LowerVector{}.cpp files. Most of the lowering patterns involve at least creating an `arith.constant`.  There are few others which involve a partial lowering to `arith` such as TestVectorGatherLowering.

Therefore, I personally feel its better to keep them seperated rather than combining two things together unless you have a much deeper reasoning why `vector.scan` and `vector.step` should be tested in a combined fashion.

> Having a dedicated TD op for such a small pattern feels like an overkill and we should probably find other patterns to group this one with. I don't really have a good suggestion just now, so I'd leave it as a TODO. 

This is what I meant as A2 isn;t it ? 
A2) include the rewriter in transform.structured.vectorize_children_and_apply_patterns

**Comment by banach-space** [comment]
> I can do A2.

+1

Please, could you also reply to this question:
> Cool, why not upstream that? That's a good optimisation and nice justification for this change.

?

**Comment by manupak** [comment]
> Cool, why not upstream that? That's a good optimisation and nice justification for this change.

We may do that but for ordering of things we need `vector.step` to survive the folder _first_  :) if that make sense

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (kuhar) [inline mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1889]: Is this valid to add RewritePattern here since it uses the dialect conversion driver?

**Review by kuhar — COMMENTED** [review]
Overall +1, we definitely don't want to expand `vector.step` for large vectors

**Review by Groverkss — APPROVED** [review]
LGTM, but please wait for @banach-space 's approval before landing

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (manupak) [reply mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1894]: yes -- it should ? AFAIK, the difference is conversion driver imposes legality post-rewrite as defined by the target.
I ve added a test to make sure which was useful in exposing the already existing rewriter pattern should be scoped to scalable vectors as per https://llvm.org/docs/LangRef.html#llvm-stepvector-intrinsic. So thanks!

File: mlir/test/Dialect/Linalg/vectorization-scalable.mlir
Comment (banach-space) [inline mlir/test/Dialect/Linalg/vectorization-scalable.mlir:186]: This is a fairly substantial change in the expected output and the reason behind that should be documented.

From what I can tell, this is due to the lowering of `vector.step` being removed from the canonicalizer. So even though the canonicalization is being run for this test, it no longer matters that match.

**Ask 1:** Please add a note in the summary/commit msg and remove the TD logic that applies canonicalizer (I expected that the output won't change).

In fact, we should remove canonicalization from this test to make this more consistent with other tests in this file (that rely on `transform.structured.vectorize` alone).

**Ask 2:** Remove canonicalization from this test.

Once you remove canonicalization, the output will become a bit uglier. My suggestion is to simplify this test. It's meant to exercise scalable vectorization, so it's fine to skip fixed-width parts. Here's a suggested simplified version
```mlir
#map = affine_map<(d0) -> (d0)>
func.func @vectorize_linalg_index(%src: tensor<?xf32>, %dest: tensor<?xf32>) -> tensor<?xf32> {
  %res = linalg.generic {
    indexing_maps = [#map],
    iterator_types = ["parallel"]
  } outs(%dest : tensor<?xf32>) {
  ^bb0(%in: f32):
    %idx = linalg.index 0 : index
    %out = tensor.extract %src[%idx] : tensor<?xf32>
    linalg.yield %out : f32
  } -> tensor<?xf32>
  return %res : tensor<?xf32>
}

module attributes {transform.with_named_sequence} {
  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
    %0 = transform.structured.match ops{["linalg.generic"]} in %arg1 : (!transform.any_op) -> !transform.any_op
    transform.structured.vectorize %0 vector_sizes [[4]] {vectorize_nd_extract} : !transform.any_op
    transform.yield
  }
}
```

**Ask 3:** Update the test.

**Comment by manupak** [comment]
@banach-space when you have some time PTAL. thank you!

**Review by banach-space — COMMENTED** [review]
> We may do that but for ordering of things we need vector.step to survive the folder first :)

Right, so the use-case remains hypothetical? ;-)

I've left some suggestions re the updated test. The comment is long, but that's me trying to save you work 😅 It's borderline out-of-scope, but why not take this refactor one step further?

File: mlir/test/Dialect/Linalg/vectorization-scalable.mlir
Comment (manupak) [reply mlir/test/Dialect/Linalg/vectorization-scalable.mlir]: I ve done three Asks except I kept `tiling_canonicalization` in the test you mentioned as it is not subject to changes in the PR.

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (kuhar) [reply mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1894]: @matthias-springer could you confirm this? I never remember what the rules are.

File: mlir/test/Dialect/Linalg/vectorization-scalable.mlir
Comment (banach-space) [inline mlir/test/Dialect/Linalg/vectorization-scalable.mlir:182]: All of the canonicalization can be removed, right?

Comment (manupak) [reply mlir/test/Dialect/Linalg/vectorization-scalable.mlir:182]: As I said I did not remove `tiling_canonicalization` as it changes the test beyond the scope of the PR.

Comment (manupak) [reply mlir/test/Dialect/Linalg/vectorization-scalable.mlir:182]: (Prior to this PR `canonicalization` is folding the vector.step . So I dont see why we have to remove this here and now)

Comment (banach-space) [reply mlir/test/Dialect/Linalg/vectorization-scalable.mlir:182]: Removing the canonicalization is not required for this change, true. But you have already re-written the test and this additional change could save us having to send/review another PR. Either way it's noise, either here or as as a new PR.

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

Please resolve the thread on rewrite patterns started by @kuhar before landing this (I have "un-resolved" it - looks like it's still ongoing).

**Comment by manupak** [comment]
Thanks @banach-space!

@kuhar if you have any suggestions lmk 
(I personally and the test shows that its fine but... I can still move to non-conversion driver)

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (banach-space) [reply mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1894]: Strictly speaking, this is lowering "Vector -> Arith" and not "Vector -> LLVM". What happens if you don't include this here?

IIUC, the whole point of this PR is _not to_ run this pattern auto-magically as part of some larger pipeline.

**Comment by kuhar** [comment]
No blockers for me, just pointed out something we should be confident about instead of guessing. And I am personally not, so just flagging.

**Comment by manupak** [comment]
I hope the test that I added inspires the confidence :)

**Comment by manupak** [comment]
PS : I dont have write access... so I d appreciate if someone can land this if there are no further concerns.

**Pull request merged**
